Resize RBD volumes on resizer request

humblec opened the pull request:
Fix # https://github.com/ceph/ceph-csi/issues/101

This patch allows RBD volumes to be expanded on demand from CO resize call. The CSI external-resizer is a sidecar container that watches the Kubernetes API server for PersistentVolumeClaim updates and triggers ControllerExpandVolume operations against a CSI endpoint if user requested more storage on PersistentVolumeClaim object.  Filesystem and block mode of RBD is supported for this expansion call. Both xfs and ext4 based filesystem expansion has been tested with this patch.


Followup/pending PR # README update

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

dillaman commented on pkg/rbd/controllerserver.go:
```diff
@@ -462,6 +462,48 @@ func storeSnapshotMetadata(rbdSnap *rbdSnapshot, cp util.CachePersister) error {
+	}
+
+	//resize volume if required
+	if (rbdVolume.VolSize) < capRange.CapacityBytes {
```
Is ```VolSize``` stored in MiB? I ask since it's treated like it's stored in bytes here but treated like it's stored in MiB on line 568

humblec replied:
You are correct, its on MiB. I will convert the CapacityBytes to MiB.

humblec replied:
Both are int64 @dillaman , so we should be good here.

dillaman commented on pkg/rbd/nodeserver.go:
```diff
@@ -102,6 +103,21 @@ func (ns *NodeServer) NodePublishVolume(ctx context.Context, req *csi.NodePublis
+
+	//TODO: rework on the path
+	resizer := resizefs.NewResizeFs(diskMounter)
+	ok., err :=  resizer.Resize(req.getVolumeId(), req.getVolumePath())
```
Nit: only if the PVC is in FS mode (i.e. don't attempt to expand a non-existent FS on a block PVC)?

humblec replied:
Yep, Have to consider both file/block modes. I have put this PR in WIP mainly to expand on this function. Will be on it soon.

humblec replied:
Done

mlmhl commented:
Hey @humblec  @dillaman Do we have any progress on the PR?

humblec commented:
@mlmhl will be updated soon. I was waiting for some PRs to go in first.

mlmhl commented:
@humblec Thanks!

humblec commented:
@Madhu-1 PTAL.

dillaman commented on pkg/rbd/nodeserver.go:
```diff
@@ -303,3 +304,18 @@ func parseFindMntResolveSource(out string) (string, error) {
 	return "", fmt.Errorf("parseFindMntResolveSource: %s doesn't match to any expected findMnt output", out)
 }
+
+func (ns *NodeServer) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolumeRequest) (*csi.NodeExpandVolumeResponse, error) {
```
This should only be valid for filesystem PVs (and don't you need to advertise ```node_expansion_required``` in ```ControllerExpandVolumeResponse``` and the ```EXPAND_VOLUME``` capability in the the node service?

Madhu-1 replied:
is this possible with spec 1.1.0?

humblec replied:
As @madhu-1 mentioned, its not possible with 1.1.0 spec. This is based on 1.1.0 implementation.

dillaman commented on pkg/rbd/controllerserver.go:
```diff
@@ -516,3 +516,46 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
+
+	return &csi.ControllerExpandVolumeResponse{
+		CapacityBytes:         req.GetCapacityRange().GetRequiredBytes(),
+		NodeExpansionRequired: true,
```
Only true if not in block mode.

XiaYinchang commented:
Any progress for resizing? Or any help needed?

mykaul commented:
> Any progress for resizing? Or any help needed?

Help is always welcome - move this PR forward, with replying to review comments, etc.

humblec commented:
@XiaYinchang this PR rebase , resolving merge conflicts ..etc are completed last week. Final testing is going on with both RWO and RWX modes. This PR will be refreshed soon.

D1abloRUS commented:
please approve this

Madhu-1 commented:
@D1abloRUS we are still reviewing the PR

Madhu-1 commented on pkg/rbd/controllerserver.go:
```diff
@@ -643,3 +643,56 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
+	rbdVol := &rbdVolume{}
+	err = genVolFromVolID(ctx, rbdVol, volID, cr)
+	if err != nil {
+		return nil, status.Errorf(codes.NotFound, "Source Volume ID %s cannot found", volID)
```
```
if err != nil {
if _, ok := err.(ErrImageNotFound); ok {
	return nil, status.Errorf(codes.NotFound, "volume ID %s not found", volID)
	}
return nil, status.Errorf(codes.Internal, err.Error())
}
```

humblec replied:
Done.

Madhu-1 commented on pkg/rbd/controllerserver.go:
```diff
@@ -643,3 +643,56 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
+
+	// always round up the request size in bytes to the nearest MiB/GiB
+	volSize := util.RoundOffBytes(req.GetCapacityRange().GetRequiredBytes())
+
```
we are calling `util.RoundOffVolSize(volSize)` two times in this code, please assign `util.RoundOffVolSize(volSize)` to a variable and use it

humblec replied:
Sure. Done.

Madhu-1 commented on pkg/rbd/controllerserver.go:
```diff
@@ -643,3 +643,56 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
+	}
+
+	nodeExpansion := true
+	return &csi.ControllerExpandVolumeResponse{
```
I think if the request size is less than the current size we are returning success but I that can we need to return the actual volume size instead of the request  size in response

humblec replied:
I think CO will have the check upfront.

Madhu-1 replied:
if CO can handle it we are good.

humblec replied:
Done.

Madhu-1 commented on pkg/rbd/rbd_util.go:
```diff
@@ -813,3 +813,22 @@ func cleanupRBDImageMetadataStash(path string) error {
+	image := rbdVol.RbdImageName
+	volSzMiB := fmt.Sprintf("%dM", rbdVol.VolSize)
+
+	klog.V(4).Infof(util.Log(ctx, "rbd: resize %s size %s using mon %s, pool %s "), image, volSzMiB, mon, rbdVol.Pool)
```
extra space in logging after `pool %s`

humblec replied:
cleared.

Madhu-1 commented on pkg/rbd/nodeserver.go:
```diff
@@ -538,6 +539,48 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 
+func (ns *NodeServer) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolumeRequest) (*csi.NodeExpandVolumeResponse, error) {
+
+	volumeID := req.GetVolumeId()
```
question: do we need to validate volumeID?

humblec replied:
may be not required, but should be ok too.

ShyamsundarR replied:
This additional check maybe required, as at present we have no other checks in this code that makes sure we are dealing with an RBD VolID and thus volume. Any plugins volume can be resized with this request, and prevention from erroneous CO requests will help avoid issues.

humblec replied:
ID validation is done, as there is no secret in this request we can not connect to the backend and validate it. More or less, this may be a rare or corner case which can be checked later too.

humblec commented:
@Madhu-1 addressed the review comments. Please check .. Thanks.

Madhu-1 commented:
@humblec please squash  the commits

humblec commented:
@Madhu-1 done. Thanks!

humblec commented:
This feature is part of https://github.com/ceph/ceph-csi/issues/557

ShyamsundarR commented on pkg/rbd/nodeserver.go:
```diff
@@ -538,6 +539,48 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
+		return nil, fmt.Errorf("rbd: resize failed on path %s, error: %v", req.GetVolumePath(), err)
+	}
+	return &csi.NodeExpandVolumeResponse{
+		CapacityBytes: req.GetCapacityRange().GetRequiredBytes(),
```
The request capacity_range and the actual capacity_range post expansion maybe different. We need to either inspect the FS/image and return the capacity or not return anything (which is fine as this field is OPTIONAL as per the CSI spec).

humblec replied:
As noted in the code ( TODO) , will have it later.

ShyamsundarR commented on pkg/rbd/nodeserver.go:
```diff
@@ -538,6 +539,48 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
+	if volumePath == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeExpandVolume volume path must be provided")
+	}
+
```
Lock and proceed? Ensures parallel expand requests or parallel Unpublish/Unstage requests do not hamper the resize efforts.

humblec replied:
Done.

ShyamsundarR commented on pkg/rbd/nodeserver.go:
```diff
@@ -538,6 +539,48 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
+	}, nil
+}
+
+func getDevicePath(ctx context.Context, mnt *mount.SafeFormatAndMount, stagingTargetPath string) (string, error) {
```
(nit) `stagingTargetPath` can be renamed as this can be either the staging or the publish path. Hence keeping this as `volumePath` or some equivalent would avoid assumptions when reading the code.

ShyamsundarR commented on deploy/rbd/kubernetes/v1.14+/csi-rbdplugin-provisioner.yaml:
```diff
@@ -85,6 +85,20 @@ spec:
+            - "--csi-address=$(ADDRESS)"
+            - "--v=5"
+            - "--csiTimeout=150s"
+            - "--leader-election"
```
(unrelated to this patch) We now will have 4 side cars performing leader election, and hence each sidecar can determine in isolation that one of the many available controller plugin instance are the leader for the said operation. This defeats the controller locks against volumes if different operations will reach different controller plugins. This either needs to be understood better (IOW, the problem does not exist) or we may need to work with kubernetes CSI side cars to see how best to ensure all side car leaders land on the same instance.

humblec replied:
Sure.. this need more discussion.

ShyamsundarR commented on pkg/rbd/nodeserver.go:
```diff
@@ -538,6 +539,48 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
+
+	diskMounter := &mount.SafeFormatAndMount{Interface: ns.mounter, Exec: mount.NewOsExec()}
+	volumePath = volumePath + "/" + req.GetVolumeId()
+	devicePath, err := getDevicePath(ctx, diskMounter, volumePath)
```
(thought) if we get the `staging_target_path` (which the spec seems to suggest that the CO should pass), we can inspect the stash file in there for the pool and image name, to use the `findDeviceMappingImage` function to detect the device name.

This can also serve as checks that we are dealing with rbd images and such.

With just the `volume_path` this is not feasible, as it could be the publish path of the volume.

humblec replied:
Sure, I will think about and follow up in a different PR accordingly.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -643,3 +643,59 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
+
+	volID := req.GetVolumeId()
+	if volID == "" {
+		return nil, status.Error(codes.InvalidArgument, "Volume ID cannot be empty")
```
Done

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -643,3 +643,59 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
 }
+
+// ControllerExpandVolume expand RBD Volumes on demand based on resizer request
+func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
```
Done.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -643,3 +643,59 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
+func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi.ControllerExpandVolumeRequest) (*csi.ControllerExpandVolumeResponse, error) {
+
+	if err := cs.Driver.ValidateControllerServiceRequest(csi.ControllerServiceCapability_RPC_EXPAND_VOLUME); err != nil {
+		klog.Warningf("invalid expand volume req: %v", protosanitizer.StripSecrets(req))
```
This need to be addressed in different PR as the same pattern is followed in other functions.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -643,3 +643,59 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
+	klog.V(4).Infof(util.Log(ctx, "rbd volume %s/%s size is %vMiB,resizing to %vMiB"), rbdVol.Pool, rbdVol.RbdImageName, rbdVol.VolSize, volSizeVal)
+	// resize volume if required
+	if rbdVol.VolSize < volSize {
+		rbdVol.VolSize = volSizeVal
```
Done.

ShyamsundarR replied:
(nit) rbd resize does take in bytes, I would prefer to stick to rounding this once and pass in bytes to the `resizeRbdImage` function

humblec commented:
@ShyamsundarR comments are addressed. Please have a final check.

humblec commented on pkg/rbd/controllerserver.go:
```diff
@@ -643,3 +643,67 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
+	nodeExpansion := false
+	if rbdVol.VolSize < volSize {
+		volSizeVal := util.RoundOffVolSize(volSize)
+		klog.V(4).Infof(util.Log(ctx, "rbd volume %s/%s size is %vMiB,resizing to %vMiB"), rbdVol.Pool, rbdVol.RbdImageName, rbdVol.VolSize, volSizeVal)
```
Done.

humblec commented on pkg/rbd/nodeserver.go:
```diff
@@ -538,6 +539,63 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
+	if err != nil {
+		// error ErrInvalidVolID may mean this is an 1.0.0 version volume, check for name
+		// pattern match in addition to error to ensure this is a likely v1.0.0 volume
+		if _, ok := err.(ErrInvalidVolID); !ok || !isLegacyVolumeID(volumeID) {
```
Done.

humblec commented on pkg/rbd/nodeserver.go:
```diff
@@ -538,6 +539,63 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
+
+	volumeID := req.GetVolumeId()
+	if volumeID == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeExpandVolume volume ID must be provided")
```
Done.

humblec commented:
@ShyamsundarR comments are addressed. PTAL. @Madhu-1 you too. thanks !

humblec commented on pkg/rbd/nodeserver.go:
```diff
@@ -538,6 +539,63 @@ func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
+	}
+
+	diskMounter := &mount.SafeFormatAndMount{Interface: ns.mounter, Exec: mount.NewOsExec()}
+	volumePath = volumePath + "/" + req.GetVolumeId()
```
Done.

humblec replied:
Done.

ShyamsundarR replied:
Right now the way the code is modified can possibly lead to a vulnerability. We need to ensure we are working on the Staging path, and not open the image meta stash without checking the same.

If the publish path were sent in here, and the user of the PV created a file that matched another image stash file, we would end up resizing the same.

As checks we need to ensure the passed in path is NOT a mount point (this eliminates it being the publish path), and if found to be a mount point then error out, as we cannot reach the stash file.

Further, if the passed in path is not a mount point, then we can ensure the passed in path + VolID is a mount point (IOW, the way stage the volume). This will ensure it is the staging path, and as a bonus we can also check if the path + VolID is a file/directory to determine if it was a block based or filesystem based PV that is in use, and not resize the block PV on the node.

Madhu-1 replied:
kubernetes sends the stagingPath as of now, why cant the check be added in follow up PR as an enhancement? once spec 1.2 is implemented ( the code remains same) instead of volumePath we will be using stagingPath

there is a check present code to make sure its filesystem or block

ShyamsundarR replied:
Depending on how a CO behaves is bad practice as such, what happens if this changes for some reason in Kubernetes? We will have to scramble at that point. I do not see why we need to get a patch in with a possible known issue that can be handled.

Also the ask is not an enhancement, the CSI 1.1.0 spec as such states this path can be either staging or publish.

ShyamsundarR replied:
Also, if this is an XFS formatted PV I think this will fail to resize as the passed in volumePath is not a mount point (if it is the staging path).

humblec replied:
@ShyamsundarR as @Madhu-1 I dont think this is a concern at all now. Couple of reasons behind , we really dont have to worry about COs other than kubernetes and also kubernetes being the CO here always send staging path in the volume_path. Its not going to be changed to publish path for a reason that it could cause  backward compatibility breakage. Considering all these I believe this approach is good enough, if really needed , I can place a followup PR .

ShyamsundarR replied:
Closing out the comments here,
- If volume path is the publish path, an unrelated PV can cause a resize request of another PV (say pv2) by crafting a stash file that points to pv2
    - This is not a concern as kubernetes always sends in the stage path is the current resolution
- We need to fix volume path to include the volID as otherwise XFS resize will fail [here](https://github.com/humblec/ceph-csi/blob/a25772c784e4b32be70bdc0ee848f4a7e619c685/vendor/k8s.io/kubernetes/pkg/util/resizefs/resizefs_linux.go#L57-L58).
- The detection of device [here](https://github.com/humblec/ceph-csi/blob/a25772c784e4b32be70bdc0ee848f4a7e619c685/vendor/k8s.io/kubernetes/pkg/util/resizefs/resizefs_linux.go#L40-L51) is insufficient to determine if this is a block volume, as a user of the block volume can also format the image as one of the FS types that is processed by Resize.
    - I am assuming this is also a concern for later.

We possibly need to open an issue to track all of the above, to ensure we deal with them as needed later.

humblec commented:
Thanks @Madhu-1 , @ShyamsundarR ptal.

ShyamsundarR commented:
> Thanks @Madhu-1 , @ShyamsundarR ptal.

@humblec Added one more comment that needs handling: https://github.com/ceph/ceph-csi/pull/331#discussion_r357116286

The above I assume should help handle both block vs file modes, as well as publish/stage paths.

Further to the above, here is my other ask,
- Request a more complete commit message for the PR, such that it is easy to understand in the future where we are at and what else needs to be done etc.

ShyamsundarR commented:
> > Thanks @Madhu-1 , @ShyamsundarR ptal.
> 
> @humblec Added one more comment that needs handling: [#331 (comment)](https://github.com/ceph/ceph-csi/pull/331#discussion_r357116286)
> 
> The above I assume should help handle both block vs file modes, as well as publish/stage paths.

Updated a summary comment for the above. Once XFS is fixed we can proceed with required merges, with the limitations stated.

> 
> Further to the above, here is my other ask,
> 
>     * Request a more complete commit message for the PR, such that it is easy to understand in the future where we are at and what else needs to be done etc.

Can we ensure the above and also squash commits as needed?

humblec commented:
Thanks @ShyamsundarR and @Madhu-1  . I have added the xfs test and changed the code to make it pass. Also squashed the commits.. Waiting for CI to pass.

humblec commented:
Updated commit message to an extend. Thanks.

humblec commented:
Finally the checks are passed ! Can you please bring back the approval @ShyamsundarR @Madhu-1 .. Thanks!

Madhu-1 commented:
@humblec please squash the last commit

humblec commented:
> @humblec please squash the last commit

Done.

humblec commented:
@Madhu-1 done.
